course5/ack/main.cc

```diff
@@ -0,0 +1,186 @@
+#include <boost/program_options.hpp>
+#include <boost/filesystem.hpp>
+#include <regex>
+
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <vector>
+
+
+/***
+Ack Project
+Ack allows for grepping for a regex pattern in all files in the entire directory tree.
+Add command-line parameters to select different regex styles, case in/sensibility, coloring of output etc.
+Get inspired by actual “ack”.
+
+Next steps:
+* Search stdin
+* actually write to file if desired
+* colors
+***/
+
+void find_matching_lines(const boost::filesystem::directory_entry& fs_entry, const std::regex& pattern, std::vector<std::string>& match_buffer)
```

Comment:
This is correct code, I don't have too many c++ comments, so I'll comment on the design:
1. the function does 2 things: opens file and searches for patter in opened file. You would achieve greater reusability if you split it. This way you could search for patterns in ifstream that's not a disk file, for instance STDIO.
2. The canonical way of returning multiple results from an algortihm function is to consume output iterator. See std::copy or std::transform for instance